File: ifaddrs4j/source/test/java/unit/io/oddsource/java/net/ifaddrs4j/TestInterfaceBrowser.java
/*
 * Copyright © 2010-2023 OddSource Code (license@oddsource.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.oddsource.java.net.ifaddrs4j;

import static org.junit.Assert.*;

import java.util.Locale;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.Test;

public class TestInterfaceBrowser
{
    private static final String OS = System.getProperty("os.name").toLowerCase(Locale.US);

    private static final boolean IS_WINDOWS = OS.contains("win");

    private static final boolean IS_MAC = OS.contains("mac") || OS.contains("darwin");

    @Test
    public void testForEachThenIterate()
    {
        try(final var browser = new InterfaceBrowser())
        {
            final var numInterfaces1 = new AtomicInteger();
            final var result = browser.forEachInterface(anInterface -> {
                numInterfaces1.getAndIncrement();
                System.out.println(anInterface.toString());
                return true;
            });

            int numInterfaces2 = 0;
            for(var anInterface : browser)
            {
                numInterfaces2++;
                assertTrue(anInterface.getIndex() > 0);
                assertNotNull(anInterface.getName());
            }

            assertTrue(result);
            assertTrue(numInterfaces1.get() > 0);
            assertEquals(numInterfaces1.get(), numInterfaces2);
        }
    }

    @Test
    public void testIterateThenForEach()
    {
        try(final var browser = new InterfaceBrowser())
        {
            int numInterfaces1 = 0;
            for(var anInterface : browser)
            {
                numInterfaces1++;
                assertTrue(anInterface.getIndex() > 0);
                assertNotNull(anInterface.getName());
            }

            final var numInterfaces2 = new AtomicInteger();
            final var result = browser.forEachInterface(ignored -> {
                numInterfaces2.getAndIncrement();
                return true;
            });

            assertTrue(result);
            assertTrue(numInterfaces1 > 0);
            assertEquals(numInterfaces1, numInterfaces2.get());
        }
    }

    @Test
    public void testIterateThenIterate()
    {
        try(final var browser = new InterfaceBrowser())
        {
            int numInterfaces1 = 0;
            for(var anInterface : browser)
            {
                numInterfaces1++;
                assertTrue(anInterface.getIndex() > 0);
                assertNotNull(anInterface.getName());
            }

            int numInterfaces2 = 0;
            for(var anInterface : browser)
            {
                numInterfaces2++;
                assertTrue(anInterface.getIndex() > 0);
                assertNotNull(anInterface.getName());
            }

            assertTrue(numInterfaces1 > 0);
            assertEquals(numInterfaces1, numInterfaces2);
        }
    }

    @Test
    public void testForEachWithReturningFalse()
    {
        try(final var browser = new InterfaceBrowser())
        {
            final var numInterfaces1 = new AtomicInteger();
            final var result1 = browser.forEachInterface(ignored -> {
                numInterfaces1.getAndIncrement();
                return false;
            });

            final var numInterfaces2 = new AtomicInteger();
            final var result2 = browser.forEachInterface(ignored -> numInterfaces2.getAndIncrement() == 0);

            assertFalse(result1);
            assertFalse(result2);
            assertEquals(numInterfaces1.get(), 1);
            assertEquals(numInterfaces2.get(), 2);
        }
    }

    @Test
    public void testGetByName()
    {
        try(final var browser = new InterfaceBrowser())
        {
            if(IS_MAC)
            {
                final var lo = browser.getInterface("lo0");
                assertEquals("lo0", lo.getName());
                assertTrue(lo.getIndex() > 0);

                final var en = browser.getInterface("en0");
                assertEquals("en0", en.getName());
                assertTrue(en.getIndex() > 0);

                assertNull(browser.getInterface("fooBar42"));
            }
        }
    }

    @Test
    public void testGetByIndex()
    {
        try(final var browser = new InterfaceBrowser())
        {
            if(IS_MAC)
            {
                final var if1 = browser.getInterface(1);
                assertEquals(1, if1.getIndex());
                assertNotNull(if1.getName());
                assertFalse(if1.getName().isEmpty());

                final var if2 = browser.getInterface(2);
                assertEquals(2, if2.getIndex());
                assertNotNull(if2.getName());
                assertFalse(if2.getName().isEmpty());

                assertNull(browser.getInterface(Integer.MAX_VALUE - 2));
            }
        }
    }

    @Test
    public void testGetInterfaces()
    {
        int interfacesFound = 0;
        boolean loopbackFound = false;
        boolean nonLoopbackFound = false;

        try(final var browser = new InterfaceBrowser())
        {
            final var interfaces = browser.getInterfaces();
            while(interfaces.hasMoreElements())
            {
                final var anInterface = interfaces.nextElement();
                interfacesFound++;
                if(anInterface.isFlagEnabled(InterfaceFlag.IsLoopback))
                {
                    loopbackFound = true;
                }
                else
                {
                    nonLoopbackFound = true;
                }
            }
        }
        assertTrue(
            "At least two interfaces should have been found, but " + interfacesFound + " found instead.",
            interfacesFound > 1
        );
        assertTrue("At least one loopback interface should have been found.", loopbackFound);
        assertTrue("At least one non-loopback interface should have been found.", nonLoopbackFound);
    }
}

File: ifaddrs4py/source/python/test/unit/test_interface_browser.py
from unittest import TestCase

from ifaddrs4py.browser import InterfaceBrowser
from ifaddrs4py.constants import IS_MACOS
from ifaddrs4py.interface import (
    Interface,
    InterfaceFlag,
)


class TestInterfaceBrowser(TestCase):
    def test_for_each_then_iterate(self) -> None:
        counts = [0, 0]

        def do_this(i: Interface) -> bool:
            print(f"{i}")
            counts[0] += 1
            return True

        browser = InterfaceBrowser()
        self.assertTrue(browser.for_each_interface(do_this))

        for interface in browser:
            counts[1] += 1
            self.assertIsNotNone(interface)

        self.assertEqual(counts[0], counts[1])
        self.assertTrue(counts[0] > 0)
        self.assertTrue(counts[1] > 0)

    def test_iterate_then_foreach(self) -> None:
        counts = [0, 0]

        def do_this(_: Interface) -> bool:
            counts[1] += 1
            return True

        browser = InterfaceBrowser()

        lens = [0, 0, 0, 0]

        lens[0] = len(browser)
        lens[1] = browser.length()

        for interface in browser:
            counts[0] += 1
            self.assertIsNotNone(interface)

        self.assertTrue(browser.for_each_interface(do_this))

        self.assertEqual(counts[0], counts[1])
        self.assertTrue(counts[0] > 0)
        self.assertTrue(counts[1] > 0)

        lens[2] = len(browser)
        lens[3] = browser.length()

        self.assertEqual(counts[0], lens[0])
        self.assertEqual(counts[0], lens[1])
        self.assertEqual(counts[0], lens[2])
        self.assertEqual(counts[0], lens[3])

    def test_iterate_then_iterate(self) -> None:
        counts = [0, 0]

        def do_this(_: Interface) -> bool:
            counts[0] += 1
            return True

        def do_that(_: Interface) -> bool:
            counts[1] += 1
            return True

        browser = InterfaceBrowser()
        self.assertTrue(browser.for_each_interface(do_this))
        self.assertTrue(browser.for_each_interface(do_that))

        self.assertTrue(counts[0] > 0)
        self.assertEqual(counts[0], counts[1])

    def test_for_each_with_returning_false(self) -> None:
        counts = [0, 0]

        def do_this(_: Interface) -> bool:
            counts[0] += 1
            return False

        def do_that(_: Interface) -> bool:
            counts[1] += 1
            return False if counts[1] > 1 else True

        browser = InterfaceBrowser()
        self.assertFalse(browser.for_each_interface(do_this))
        self.assertFalse(browser.for_each_interface(do_that))

        self.assertEqual(counts[0], 1)
        self.assertEqual(counts[1], 2)

    def test_get_by_name(self) -> None:
        browser = InterfaceBrowser()

        if IS_MACOS:
            lo = browser.get_interface("lo0")
            self.assertIsNotNone(lo)
            assert lo is not None  # for mypy
            self.assertEqual(lo.name, "lo0")
            self.assertTrue(lo.index > 0)

            lo = browser["lo0"]
            self.assertEqual(lo.name, "lo0")
            self.assertTrue(lo.index > 0)

            en = browser.get_interface("en0")
            self.assertIsNotNone(en)
            assert en is not None  # for mypy
            self.assertEqual(en.name, "en0")
            self.assertTrue(en.index > 0)

            en = browser["en0"]
            self.assertEqual(en.name, "en0")
            self.assertTrue(en.index > 0)

            self.assertIsNone(browser.get_interface("fooBar42"))
            self.assertRaises(KeyError, lambda: browser["fooBar42"])

    def test_get_by_index(self) -> None:
        browser = InterfaceBrowser()

        if IS_MACOS:
            if1 = browser.get_interface(1)
            self.assertIsNotNone(if1)
            assert if1 is not None  # for mypy
            self.assertEqual(if1.index, 1)
            self.assertIsNotNone(if1.name)

            if1 = browser[1]
            self.assertEqual(if1.index, 1)
            self.assertIsNotNone(if1.name)

            if2 = browser.get_interface(2)
            self.assertIsNotNone(if2)
            assert if2 is not None  # for mypy
            self.assertEqual(if2.index, 2)
            self.assertIsNotNone(if2.name)

            if2 = browser[2]
            self.assertEqual(if2.index, 2)
            self.assertIsNotNone(if2.name)

            self.assertIsNone(browser.get_interface(2_147_483_645))
            self.assertRaises(KeyError, lambda: browser[2_147_483_645])

    def test_get_interfaces(self) -> None:
        interfaces_found = 0
        loopback_found = False
        non_loopback_found = False

        browser = InterfaceBrowser()
        for interface in browser.get_interfaces():
            interfaces_found += 1
            if interface.is_flag_enabled(InterfaceFlag.IsLoopback):
                loopback_found = True
            else:
                non_loopback_found = True

        self.assertTrue(interfaces_found > 1)
        self.assertTrue(loopback_found)
        self.assertTrue(non_loopback_found)

    def test_no_segfault(self) -> None:
        counts = [0, 0, 0]

        for interface in InterfaceBrowser().get_interfaces():
            counts[0] += 1

        for interface in InterfaceBrowser():
            counts[1] += 1

        def do_this(_: Interface) -> bool:
            counts[2] += 1
            return True

        self.assertTrue(InterfaceBrowser().for_each_interface(do_this))

        self.assertTrue(counts[0] > 0)
        self.assertTrue(counts[1] > 0)
        self.assertTrue(counts[2] > 0)
        self.assertEqual(counts[0], counts[1])
        self.assertEqual(counts[1], counts[2])
